corehq/apps/sso/tests/test_backends.py

```diff
@@ -0,0 +1,225 @@
+        """
+        SsoBackend should fail to return a user if the username passed to it
+        is not a valid email address (missing email domain / no `@`). It should
+        also populate request.sso_login_error with an error message.
+        """
+        self._set_saml_session_index()
+        user = auth.authenticate(
+            request=self.request,
+            username='badusername',
+            idp_slug=self.idp.slug
+        )
+        self.assertIsNone(user)
+        self.assertEqual(
+            self.request.sso_login_error,
+            "Username badusername is not valid."
+        )
+
+    def test_login_error_if_email_domain_does_not_exist(self):
+        """
+        SsoBackend should fail to return a user if the username passed to it
+        matches an email domain that is not mapped to any Identity Provider.
+        It should also populate request.sso_login_error with an error message.
+        """
+        self._set_saml_session_index()
+        user = auth.authenticate(
+            request=self.request,
+            username='b@vwx.link',
+            idp_slug=self.idp.slug
+        )
+        self.assertIsNone(user)
+        self.assertEqual(
+            self.request.sso_login_error,
+            "The Email Domain vwx.link is not allowed to authenticate with "
+            "this Identity Provider."
+        )
+
+    def test_login_error_if_email_domain_is_not_authorized(self):
+        """
+        SsoBackend should fail to return a user if the username passed to it
+        matches an email domain that exists, but is not authorized to
+        authenticate with the Identity Provider in the request. It should also
+        populate request.sso_login_error with an error message.
+        """
+        self._set_saml_session_index()
+        AuthenticatedEmailDomain.objects.create(
+            email_domain='vwx.link',
+            identity_provider=self.idp_two,  # note that this is not self.idp
+        )
+        user = auth.authenticate(
+            request=self.request,
+            username='b@vwx.link',
+            idp_slug=self.idp.slug  # note that this is self.idp, not idp_two
+        )
+        self.assertIsNone(user)
+        self.assertEqual(
+            self.request.sso_login_error,
+            "The Email Domain vwx.link is not allowed to authenticate with "
+            "this Identity Provider."
+        )
+
+    def test_login_error_if_user_does_not_exist(self):
+        """
+        SsoBackend should fail to return a user if the username passed to does
+        not exist. It should also populate request.sso_login_error with an
+        error message.
+
+        todo this test will change with additional user creation workflows
+         that will be introduced later
+        """
+        self._set_saml_session_index()
+        AuthenticatedEmailDomain.objects.create(
+            email_domain='vaultwax.com',
+            identity_provider=self.idp,
+        )
```

Comment:
Would it make sense to move this into the setup? Like, this suite has a given domain that is always covered by a specific IDP, and then the tests that want to change that assumption verify against a different domain or provider? This feels like an unrelated implementation detail in this specific test.